utils/wxrc/wxrc.cpp

```diff
@@ -626,41 +627,105 @@ void XmlResApp::MakePackageZIP(const wxArrayString& flist)
     wxFileOffset offset = file.Length();
     wxASSERT_MSG( offset >= 0 , wxT("Invalid file length") );
 
-    const size_t lng = wx_truncate_cast(size_t, offset);
+    // [2024-03-15]
+    //
+    //   There are two environments (processes) to consider:
+    //     - the resource compiler;
+    //     - the process using the resource.
+    //
+    //   Previously, we used to check whether `offset` fitted into `size_t`.
+    //
+    //   But this used to check against the `size_t` of the resource compiler.
+    //   The check might have passed, but this might have been misleading
+    //   (if the resource compiler is a 64-bit process and the process using the resource is a 32-bit process).
+    //
+    //   Also, for Windows at least, the file format for modules (executables and shared libraries)
+    //   which are to be loaded in 64-bit processes
+    //   is not called PE64, but instead PE32+,
+    //   and yes, in this format virtual addresses are indeed 64-bit,
+    //   but in fact relative virtual addresses (and also file offsets) are 32-bit
+    //   (i.e. a single module cannot span for more than 4 GiB in virtual memory).
+    //
+    //   It follows that the check might have also been misleading
+    //   in the case of the resource compiler being a 64-bit Linux process
+    //   and the process using the resource being a Windows process -- even a 64-bit Windows process.
+    //
+    //   Therefore, we now check whether the size of the resource fits
+    //   not into `size_t`, but instead into `uint32_t`,
+    //   i.e. we do not support resources larger than 4 GiB.
+    //
+    const wxUint32 lng = wx_truncate_cast(wxUint32, offset);
     wxASSERT_MSG( static_cast<wxFileOffset>(lng) == offset,
                   wxT("Huge file not supported") );
 
     snum.Printf(wxT("%i"), num);
-    output.Printf(wxT("static size_t xml_res_size_") + snum + wxT(" = %lu;\n"),
-                  static_cast<unsigned long>(lng));
-    output += wxT("static unsigned char xml_res_file_") + snum + wxT("[] = {\n");
+
+    tmp.Printf(wxT("// File '%s':\n"), filename);
+    output += tmp;
+
+    tmp.Printf(wxT("const size_t xml_res_size_") + snum + wxT(" = %u;\n"), lng);
+    output += tmp;
+
+    output += wxT("const unsigned char xml_res_file_") + snum + wxT("[] = {\n");
     // we cannot use string literals because MSVC is dumb wannabe compiler
     // with arbitrary limitation to 2048 strings :(
 
-    unsigned char *buffer = new unsigned char[lng];
-    file.Read(buffer, lng);
+    std::vector<unsigned char> buffer(lng);
+    file.Read(buffer.data(), lng);
 
-    for (size_t i = 0, linelng = 0; i < lng; i++)
+    for (wxUint32 i = 0, linelng = 0, sectionlng = 0; ; )
     {
-        tmp.Printf(wxT("%i"), buffer[i]);
-        if (i != 0) output << wxT(',');
-        if (linelng > 70)
+        if (! i || i >= lng || sectionlng >= 0x100)
         {
-            linelng = 0;
-            output << wxT("\n");
+            if (linelng) wxFAIL;
```

Comment:
I would like to make an assertion:

```
some_azzert_macro(! linelng);
```

(Perhaps that way of expressing in the source code might have been more clear...)

Strangely enough, the intent has been to clarify for readers of the code that, despite the maybe complex logic, at this point `linelng` should be zero, otherwise the programmer has made an error.

I wanted the azzertion to always be enabled, even in `_NDEBUG` builds (i.e. in "release" builds).

If alright with you, may I ask whether a macro already exists for this, please ?